client/src/objects/drawin.rs

```diff
@@ -80,54 +182,86 @@ impl<'lua> Drawin<'lua> {
-        {
-            let mut drawin = self.state_mut()?;
-            drawin.visible = val;
+        let mut state = self.state_mut()?;
+        trace!("set_visible(): {}, old: {}", val, state.visible);
+        if val == state.visible {
+            return Ok(());
         }
+
+        state.visible = val;
+
+        drop(state);
+
         if val {
-            self.map(lua)
+            self.map(lua)?
         } else {
-            self.unmap()
+            self.unmap(lua)?
         }
+
+        trace!("will signal property::visible");
+
+        Object::emit_signal(lua, self, "property::visible", Value::Nil)?;
+
+        Ok(())
     }
 
     fn map(&mut self, lua: rlua::Context<'lua>) -> rlua::Result<()> {
         // TODO other things
+        trace!("drawin::map");
         self.update_drawing(lua)?;
         Ok(())
     }
 
-    fn unmap(&mut self) -> rlua::Result<()> {
+    fn unmap(&mut self, lua: rlua::Context<'lua>) -> rlua::Result<()> {
         // TODO?
+        trace!("drawin::unmap");
+        self.update_drawing(lua)?;
         Ok(())
     }
 
     pub fn get_geometry(&self) -> rlua::Result<Area> {
         Ok(self.state()?.geometry)
     }
 
+    /// Move and/or resize a drawin
     fn resize(&mut self, lua: rlua::Context<'lua>, geometry: Area) -> rlua::Result<()> {
-        {
+        trace!("drawin::resize");
+        let signals = {
             let mut state = self.state_mut()?;
-            let old_geometry = state.geometry;
-            state.geometry = geometry;
-            {
-                let Size {
-                    ref mut width,
-                    ref mut height
-                } = state.geometry.size;
-                if *width == 0 {
-                    *width = old_geometry.size.width;
-                }
-                if *height == 0 {
-                    *height = old_geometry.size.height
-                }
-            }
-            state.geometry_dirty = true;
-            // TODO emit signals
-            // TODO update screen workareas like in awesome? Might not be necessary
+            state.update_geometry(geometry)
+        };
+
+        if signals.len() > 0 {
+            self.update_drawing(lua)?;
+        }
+
+        // TODO update screen workareas like in awesome? Might not be necessary
+        for (signal, args) in signals {
+            trace!("drawin::resize: signaling {}", signal);
+            Object::emit_signal(lua, self, &signal, args)?;
         }
-        self.update_drawing(lua)
+
+        Ok(())
+    }
+
+    fn set_struts(&mut self, lua: rlua::Context<'lua>, struts: Margin) -> rlua::Result<()> {
+        trace!("set_struts({:?})", struts);
+        if struts == self.state()?.struts {
+            return Ok(());
+        }
+
+        self.state_mut()?.struts = struts;
+
+        // TODO(ried): emit screen change workarea event?
+        // TODO(ried): remove since awesome does not emit this event?
```

Comment:
If awesome doesn't emit the event then neither should we. Later once we are compatible we can see about potentially fixing these mistakes